python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,473 @@
+                nvtext.ngrams_tokenize(tokens, N=n, sep=' '))
+
+        return ngrams
+
+    def _char_ngrams(self, text_document):
+        """Tokenize text_document into a sequence of character n-grams"""
+        # tokens = nvtext.character_tokenize(text_document)
+        # return self._word_ngrams(tokens)
+        raise NotImplementedError("Character-level ngrams is not yet"
+                                  " supported by cuML.")
+
+    def _char_wb_ngrams(self, text_document):
+        """Whitespace sensitive char-n-gram tokenization.
+        Tokenize text_document into a sequence of character n-grams
+        operating only inside word boundaries. n-grams at the edges
+        of words are padded with space."""
+        text_document = self._surround_with_space(text_document)
+        return self._char_ngrams(text_document)
+
+    @staticmethod
+    def _surround_with_space(nvstr):
+        """Adds spaces before and after each strings of a nvstrings"""
+        s = len(nvstr)
+        spaces = Series('').repeat(s)._column.nvstrings
+        return spaces.cat(nvstr.cat(spaces, sep=' '), sep=' ')
+
+    def build_preprocessor(self):
+        """Return a function to preprocess the text before tokenization.
+        Returns
+        -------
+        preprocessor: callable
+              A function to preprocess the text before tokenization.
+        """
+        if self.preprocessor is not None:
+            return self.preprocessor
+        return partial(_preprocess, lower=self.lowercase,
+                       remove_punctuation=self.remove_punctuation)
+
+    def get_stop_words(self):
+        """Build or fetch the effective stop words list.
+        Returns
+        -------
+        stop_words: list or None
+                A list of stop words.
+        """
+        if self.stop_words == "english":
+            return list(ENGLISH_STOP_WORDS)
+        elif isinstance(self.stop_words, str):
+            raise ValueError("not a built-in stop list: %s" % self.stop_words)
+        elif self.stop_words is None:
+            return None
+        else:  # assume it's a collection
+            return list(self.stop_words)
+
+    def _build_ngram_vocab(self, docs):
+        """
+        Build vocabulary when using ngrams.
+
+        Returns
+        -------
+        ngrams: nvstrings
+            The vocabulary of n-grams.
+        """
+        if self.analyzer == 'char':
+            return self._char_ngrams(docs)
+        elif self.analyzer == 'char_wb':
+            return self._char_wb_ngrams(docs)
+        elif self.analyzer == 'word':
+            return self._word_ngrams(docs)
+        else:
+            raise ValueError('%s is not a valid tokenization scheme/analyzer' %
+                             self.analyzer)
+
+    def _remove_stop_words(self, vocab):
+        stop_words = self.get_stop_words()
+        vocab = self._surround_with_space(vocab)
+        for w in (f' {w} ' for w in stop_words):
```

Comment:
We should use `replace_tokens` instead of looping here as this will be expensive as stop words increase in count. 

https://rapidsai.github.io/projects/nvstrings/en/0.9.0/api.html#nvtext.replace_tokens

Reply:
Indeed! I didn't know about `replace_tokens` and I was looking for workarounds but you are right I think it's exactly what we want here.

Reply:
While I think this is a better approach indeed, I found a small caveat and I'm not sure what's the best way to go about it.

The replace_tokens needs additional steps in the case of ngrams. 
For instance, this input `["and me too andy andy too"]` will generate those bigrams: 
`['and me', 'andy andy', 'andy too', 'me too', 'too andy']`. After `replace_tokens`, the ngrams becomes `['   ', 'andy andy', 'andy  ', '   ', '  andy']`. We can then remove the strings with only spaces, but we also want to remove `'andy  '` and `'  andy'`  because those ngrams contained stop words (respectively `'andy too'`  and `'too andy'`).

- One way I can think of would be to use a special marker, using `replace_tokens(vocab, stop_words, marker)` and then removing every string that contains that special marker. But of course this is not ideal since the marker could always happen to be in the input documents. We could mitigate this by providing a parameter for the marker.
- Second possibility is to remove stop words from the input document instead of the vocabulary. This can be done using only `replace_tokens(docs, stop_words, delimiter)`. I like this approach better but I'm worried this might introduce more discrepancy between our implementation and sklearn's.

Do you have a preferred approach?

Reply:
> One way I can think of would be to use a special marker, using replace_tokens(vocab, stop_words, marker) and then removing every string that contains that special marker. But of course this is not ideal since the marker could always happen to be in the input documents. We could mitigate this by providing a parameter for the marker.

I think we can add a special market as i think that can be a safe assumption and just add a check to throw an error if input contains that character, i feel this is  a rare enough problem not to matter. 

> Second possibility is to remove stop words from the input document instead of the vocabulary. This can be done using only replace_tokens(docs, stop_words, delimiter). I like this approach better but I'm worried this might introduce more discrepancy between our implementation and sklearn's.

When you discrepancy can you add more info around what particular ones and if you think that output will change ?  Like you i prefer this approach as that has better performance characteristics.

Reply:
> I think we can add a special market as i think that can be a safe assumption and just add a check to throw an error if input contains that character, i feel this is a rare enough problem not to matter.
Fair point. We can do that.

Reply:
> When you discrepancy can you add more info around what particular ones and if you think that output will change ? Like you i prefer this approach as that has better performance characteristics.

I don't have a specific example of a different output, but since sklearn's implementation deliberately chose to remove the stop words at the vocabulary step, I'm thinking doing otherwise might introduce bugs.

I think for now we can use this second solution as it is more efficient, but let's keep an eye open for discrepancy on this side.

For reference, in case we were to chose the first solution, here is the corresponding implementation of `_remove_stop_words` that I wrote while testing:
```python
    def _remove_stop_words(self, vocab):
        marker = '!@#$'

        stop_words = Series(self._get_stop_words())._column.nvstrings

        vocab = nvtext.replace_tokens(vocab, stop_words, marker)

        stop_words_mask = cp.empty(shape=len(vocab), dtype=cp.bool)
        vocab.contains(marker, regex=False, devptr=stop_words_mask.data.ptr)

        idx = cp.where(stop_words_mask)[0].astype(cp.int32)
        vocab = vocab.remove_strings(idx.data.ptr, len(idx))

        return vocab
```

Reply:
Agreed, lets keep an eye out for bugs , might be worth adding a comment stating this assumption.